src/commands/classified/external.rs

```diff
@@ -50,158 +49,243 @@ impl Decoder for LinesCodec {
-            } else {
-                process = process.arg(arg.as_ref());
+    let popen = process.popen();
+    if let Ok(mut popen) = popen {
+        let _ = popen.detach();
+        if is_last {
+            while let None = popen.poll() {
+                std::thread::sleep(std::time::Duration::new(0, 100000000));
             }
+
+            let _ = popen.terminate();
+            Ok(InputStream::empty())
+        } else {
+            let stdout = popen.stdout.take().unwrap();
+            let file = futures::io::AllowStdIo::new(stdout);
+            let stream = Framed::new(file, LinesCodec {});
+            let stream = stream.map(move |line| line.unwrap().into_value(&name_tag));
+            let stream = stream.boxed() as BoxStream<'static, Value>;
+            Ok(stream.into())
         }
+    } else {
+        Err(ShellError::labeled_error(
+            "Command not found",
+            "command not found",
+            name_tag,
+        ))
     }
+}
 
-    process = process.cwd(context.shell_manager.path());
+async fn run_with_stdin(
+    command: ExternalCommand,
+    context: &mut Context,
+    mut input: InputStream,
+    is_last: bool,
+) -> Result<InputStream, ShellError> {
+    let name_tag = command.name_tag.clone();
+    let home_dir = dirs::home_dir();
 
-    trace!(target: "nu::run::external", "cwd = {:?}", context.shell_manager.path());
+    let mut process = Exec::cmd(&command.name);
+    for arg in command.args.iter() {
+        // Let's also replace ~ as we shell out
+        let arg = shellexpand::tilde_with_context(arg.deref(), || home_dir.as_ref());
 
-    let mut process = match stream_next {
-        StreamNext::Last => process,
-        StreamNext::External | StreamNext::Internal => {
-            process.stdout(subprocess::Redirection::Pipe)
+        let arg_chars: Vec<_> = arg.chars().collect();
+        if (arg_chars.len() > 1 && arg_chars[0] == '"' && arg_chars[arg_chars.len() - 1] == '"')
+            || (arg_chars.len() > 1
+                && arg_chars[0] == '\''
+                && arg_chars[arg_chars.len() - 1] == '\'')
+        {
+            // quoted string
+            let new_arg: String = arg_chars[1..arg_chars.len() - 1].iter().collect();
+            process = process.arg(new_arg);
+        } else {
+            process = process.arg(arg.as_ref());
         }
-    };
+    }
 
-    trace!(target: "nu::run::external", "set up stdout pipe");
+    process = process.cwd(context.shell_manager.path());
+    trace!(target: "nu::run::external", "cwd = {:?}", context.shell_manager.path());
 
-    if let Some(stdin) = stdin {
-        process = process.stdin(stdin);
+    if !is_last {
+        process = process.stdout(subprocess::Redirection::Pipe);
+        trace!(target: "nu::run::external", "set up stdout pipe");
     }
 
+    // NOTE can't use a subprocess::make_pipe here. There seems to be no way to close stdin
+    //      after doing so, which makes certain commands block forever (e.g., cat).
```

Comment:
Interesting... line 200 below also seems good. Does make_pipe offer a different advantage?

Reply:
I think the one advantage of `make_pipe` would actually be not here, but for the case where we have `$it`. In that case we can have a single `std::fs::File` to read from, and have all the commands we spawn off use the other end of the pipe. The advantage there could be hypothetical and unnecessary 😬